File: backend/app/utils/id_card.py
"""身份证号校验与性别提取工具。"""

# 加权因子
WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
CHECK_CODES = "10X98765432"


def validate_id_card(id_card: str) -> bool:
    """校验 18 位身份证号是否合法。"""
    if len(id_card) != 18:
        return False
    body = id_card[:17]
    if not body.isdigit():
        return False
    check = id_card[17].upper()
    total = sum(int(body[i]) * WEIGHTS[i] for i in range(17))
    expected = CHECK_CODES[total % 11]
    return check == expected


def extract_gender(id_card: str) -> str:
    """从身份证号提取性别：倒数第二位奇数为男，偶数为女。"""
    return "male" if int(id_card[16]) % 2 == 1 else "female"

File: frontend/src/utils/id-card.js
const WEIGHTS = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
const CHECK_CODES = '10X98765432'

export function validateIdCard(idCard) {
  if (!idCard || idCard.length !== 18) return false
  const body = idCard.substring(0, 17)
  if (!/^\d{17}$/.test(body)) return false
  const check = idCard[17].toUpperCase()
  let total = 0
  for (let i = 0; i < 17; i++) {
    total += parseInt(body[i]) * WEIGHTS[i]
  }
  return check === CHECK_CODES[total % 11]
}

export function extractGender(idCard) {
  if (!idCard || idCard.length < 17) return ''
  return parseInt(idCard[16]) % 2 === 1 ? 'male' : 'female'
}
